ksql-engine/src/main/java/io/confluent/ksql/util/StatementUtil.java

```diff
@@ -0,0 +1,141 @@
+  }
+
+  public static <T extends Statement> PreparedStatement<T> withInferredSchema(
+      final T statement,
+      final String statementText,
+      final SchemaRegistryClient schemaRegistryClient) {
+    return withInferredSchema(PreparedStatement.of(statementText, statement), schemaRegistryClient);
+  }
+
+  @SuppressWarnings("unchecked")
+  public static <T extends Statement> PreparedStatement<T> withInferredSchema(
+      final PreparedStatement<T> statement,
+      final SchemaRegistryClient schemaRegistryClient
+  ) {
+    if (!(statement.getStatement() instanceof AbstractStreamCreateStatement)) {
+      return statement;
+    }
+
+    final AbstractStreamCreateStatement statementWithSchema = AvroUtil.checkAndSetAvroSchema(
+        (AbstractStreamCreateStatement) statement.getStatement(), schemaRegistryClient);
+
+    if (statementWithSchema == statement.getStatement()) {
+      return statement;
+    }
+
+    return (PreparedStatement) PreparedStatement.of(
+        SqlFormatter.formatSql(statementWithSchema),
+        statementWithSchema);
+  }
+
+  /**
+   * Resolves all topic properties from the statement so that topic name, number partitions
+   * and number of replicas are explicitly specified.
+   *
+   * @return a prepared statement with the resolved properties and updated statement text
+   */
+  public static PreparedStatement withInferredSinkTopic(
+      final PreparedStatement statement,
+      final Map<String, Object> propertyOverrides,
+      final KsqlConfig ksqlConfig
+  ) {
+    if (!(statement.getStatement() instanceof CreateAsSelect)) {
+      return statement;
+    }
+
+    final CreateAsSelect cas = (CreateAsSelect) statement.getStatement();
+
+    final String topic = topicName(cas, ksqlConfig);
+    final int partitions = numPartitions(cas, ksqlConfig, propertyOverrides);
+    final short replicas = numReplicas(cas, ksqlConfig, propertyOverrides);
+
+    final Map<String, Expression> properties = new HashMap<>(cas.getProperties());
+    properties.put(DdlConfig.KAFKA_TOPIC_NAME_PROPERTY, new StringLiteral(topic));
+    properties.put(KsqlConstants.SINK_NUMBER_OF_REPLICAS, new IntegerLiteral(replicas));
+    properties.put(KsqlConstants.SINK_NUMBER_OF_PARTITIONS, new IntegerLiteral(partitions));
+
+    final CreateAsSelect withTopic = cas.copyWith(properties);
+    final String withTopicText = SqlFormatter.formatSql(withTopic) + ";";
+
+    return PreparedStatement.of(withTopicText, withTopic);
+  }
+
+  /**
+   * If the topic name is specified in the request, use that topic name directly, otherwise
+   * construct it using the default topic prefix and the name of the output source.
+   */
+  private static String topicName(final CreateAsSelect cas, final KsqlConfig cfg) {
+    final Expression topicProperty = cas.getProperties().get(DdlConfig.KAFKA_TOPIC_NAME_PROPERTY);
+    return (topicProperty == null)
+        ? cfg.getString(KsqlConfig.KSQL_OUTPUT_TOPIC_NAME_PREFIX_CONFIG) + cas.getName().getSuffix()
```

Comment:
Does `getName` return the name in upper case?

Reply:
Nope, but I believe some other part of the stack that is called before this does.

Reply:
I'm not seeing where that happens. CreateAsSelect gets created by AstBuilder, which just uses the string from the statement. We should be sure we create sink topics with upper case names, or call out the change in behavior in the docs (which we'll probably have to do anyway since we're changing the way these are computed as part of the topic management design)

Reply:
It looks like you're right - I thought I had tested this out, but clearly not. Good catch! I will update this to call `toUpperCase`

>since we're changing the way these are computed as part of the topic management design

This is no longer a requirement for the MVP. We wanted to prepend the queryID to the topic name so that we make sure that we don't delete a topic that we created in the past with the same name. With server-side deletes this doesn't happen anymore - either I can't create the topic because the topic with the same name already exists or it's already deleted (in which case I can safely delete the topic I just created). The only race that it solves (which I don't think we should worry about) is the case where two different users are interacting with two different servers issuing commands at the same time for the same topic name:
```
Server A:   |---CREATE X---|                              |---DELETE X---|
Server B:                  |---DELETE X---||---CREATE X---|
```
To me, this falls in the category of "inter-server" communication.

Reply:
Nevermind. I found where it happens:
```java
  public static String getIdentifierText(final SqlBaseParser.IdentifierContext context) {
    if (context instanceof SqlBaseParser.QuotedIdentifierAlternativeContext) {
      return unquote(context.getText(), "\"");
    } else if (context instanceof SqlBaseParser.BackQuotedIdentifierContext) {
      return unquote(context.getText(), "`");
    } else {
      return context.getText().toUpperCase();
    }
  }
```
In `getQualifiedName` in `AstBuilder`.